### PR TITLE
accelerate `binary_cross_entropy_with_logits` by using `log_sigmoid` operator

### PR DESCRIPTION
**Pull request by qingpeng9802**
When I was reimplementing BCEwithLogits, I found that `log_sigmoid` operator could accelerate the function.

Simple benchmark on AMD 3600 CPU Ubuntu 22.04:
|avg time (ms)|with `pos_weight`|no `pos_weight`|
|-|-|-|
|original|1986|1658|
|this PR|1295|995|

faster 35-40%. This is probably benefited by the `log_sigmoid` vectorization code.

CUDA benchmark was not obtained, but I believe CUDA can be also benefited by reduecing kernel launches as https://github.com/pytorch/pytorch/pull/11054#issuecomment-442233714 and https://github.com/pytorch/pytorch/pull/78267#issue-1248398454 mentioned.

The simple benchmark cpp file:
[demo.txt](https://github.com/pytorch/pytorch/files/13635355/demo.txt)

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Might be related: https://github.com/pytorch/pytorch/issues/9141 https://github.com/goodlux/pytorch/commit/f6b1259459eb461f780c4c515f50f5b3ec3e0921#r134742895

**Comment by qingpeng9802** [comment]
Hi @zou3519 Could you please help approve the CI workflows so I can check the tests and use the artifacts?

**Comment by qingpeng9802** [comment]
Hi @jbschlosser Not sure if you are the right guy to review this PR, but could you please help open the CI workflow so I can check the tests and use the artifacts?

**Comment by qingpeng9802** [comment]
cc @jbschlosser @zou3519 is there anyone related can help open the CI workflow so I can check the tests and use the artifacts?

**Comment by qingpeng9802** [comment]
cc @jbschlosser @zou3519 is there anyone related can help open the CI workflow so I can check the tests and use the artifacts?

**Comment by qingpeng9802** [comment]
cc @lezcano @mikaylagawarecki @jbschlosser @bdhirsh @zou3519 
Not sure who should review this, but can someone help open CI so I can test the artifacts?

**Review by lezcano — COMMENTED** [review]
We can even use `log_sigmoid_out(input, input)` to perform the operation in place and avoid allocating a new tensor.

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -359,21 +360,20 @@ Tensor binary_cross_entropy_with_logits(const Tensor& input, const Tensor& targe
...
-        auto log_weight = (pos_weight - 1).mul(target).add_(1);
-        loss = (1 - target).mul_(input).add_(log_weight.mul_(((-max_val).exp_().add_((-input - max_val).exp_())).log_().add_(max_val)));
-    } else {
-        loss = (1 - target).mul_(input).add_(max_val).add_((-max_val).exp_().add_((-input -max_val).exp_()).log_());
-    }
+  Tensor loss;
+  if (pos_weight.defined()) {
+      // pos_weight need to be broadcasted, thus mul(target) is not inplace.
+      auto log_weight = (pos_weight - 1).mul(target).add_(1);
+      loss = (1 - target).mul_(input).add_(log_weight.mul_(at::log_sigmoid(input).neg_()));
```
Comment (lezcano) [inline aten/src/ATen/native/Loss.cpp:367]: You can also avoid the `neg_` by swapping the `add_` by a `sub_`. Same below.

**Comment by qingpeng9802** [comment]
> We can even use `log_sigmoid_out(input, input)` to perform the operation in place and avoid allocating a new tensor.

@lezcano I was thinking of using this `out` op, but I was not sure if it is safe to perform in-place op on the `input` because it seems that the `input` should not be modified (as a convention?). Just a double check.

**Comment by lezcano** [comment]
Ah, right, sorry, no, the input should not be modified. That op should be out-of-place :)

**Review by lezcano — DISMISSED** [review]
Cool, thanks!

**Comment by qingpeng9802** [comment]
Not quite understand the failed case, but I guess this code should has the same copy as this implementation?

https://github.com/pytorch/pytorch/blob/4712c7dac802917e74c2ad6c5bb94924220e59e2/torch/_decomp/decompositions.py#L3807-L3828

**Comment by lezcano** [comment]
The issue here is, why is that implementation giving such different results when compared to the implementation with log_sigmoid?

**Comment by qingpeng9802** [comment]
> The issue here is, why is that implementation giving such different results when compared to the implementation with log_sigmoid? Are the two really equivalent?

`RuntimeError: Difference from float64 is larger with decomposition binary_cross_entropy_with_logits.default than original on output 0. Original max diff: 0.0008904983620294793, Decomp max diff: 0.0010626266379705207`

My understanding is that this warning is raised for the *decomposition implementation* because the *decomposition implementation* has a larger diff (like https://github.com/pytorch/pytorch/pull/80964#issuecomment-1190660251 and https://github.com/pytorch/pytorch/pull/115306#issue-2029549456), which means that this *PR implementation* reach a better precision. Thus, it is safe to change the *decomposition implementation*.

Any comment about this? @lezcano

**Comment by lezcano** [comment]
Ah, right, I had misread the error. `0.0009` is quite reasnoable, and you assessment is correct. This surely comes from using `log1p`, which is more exact. Changing the decomposition SGTM.

**Comment by qingpeng9802** [comment]
It looks like a record matching issue. All records with `binary_cross_entropy_with_logits` are triggered.

https://github.com/pytorch/pytorch/blob/1967165d4db648ab2a61afd99cbeb6f260e35ea8/test/profiler/test_memory_profiler.py#L1150

Simply updating the records should work, but I am not sure if this is safe.

**Comment by lezcano** [comment]
yep, as it says in the test, just run those tests with `EXPECTTEST_ACCEPT=1` and commit the results

**Comment by DanilBaibak** [comment]
Hi @qingpeng9802! Sorry, I have to revert your PR because it leads to internal failures. We are working to resolve the problem. As soon as we have some suggestions, I will get back to you.If the problem not with your changes, I will re-import your PR.

**Comment by malfet** [comment]
@DanilBaibak this looks like a good-old mobile model retracing, did you run it on the internal diff?

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
@lezcano sorry, yes I think it could be landed

**Pull request closed without merging**